### PR TITLE
[OP-40] makeDataSet refactored

### PR DESCRIPTION
**Pull request by QFer**
* refactored makeDataSet1D, makeDataSet1DPlain, makeDataSet2D, makeDataSet2DPlain
* combined generic functionality
* created 100% unit tests for these methods

### CODE REVIEW

File: src/qtt/data.py [hunk 1]
```diff
@@ -260,7 +261,8 @@ def add_comment(txt, dataset=None, verbose=0):
                 pass
         else:
             raise NotImplementedError('dataset not specified and _latest_datasets not available')
-            dataset = qcodes.DataSet._latest
+            # unreachable code
```
Comment (CumulonimbusCalvus) [inline src/qtt/data.py]: Why not remove the unreachable code?

File: src/qtt/data.py [hunk 2]
```diff
@@ -1029,183 +1031,326 @@ def saveExperimentData(outputdir, dataset, tag, dstr):
     write_data(path, dataset)
 
 
-def makeDataSet1Dplain(xname, x, yname, y=None, xunit=None, yunit=None, location=None, loc_record=None):
-    ''' Make DataSet with one 1D array and one setpoint array
+def determine_parameter_unit(parameter):
+    """ Determine unit associated with a qcodes.parameter, otherwise None """
```
Comment (CumulonimbusCalvus) [inline src/qtt/data.py]: No args documentation?

File: src/qtt/data.py [hunk 3]
```diff
@@ -1029,183 +1031,326 @@ def saveExperimentData(outputdir, dataset, tag, dstr):
...
+    if isinstance(parameter, qcodes.Parameter):
+        return parameter.unit
+    else:
+        return None
+
+
+def _check_parameter(p):
+    """ Check p to be of correct type - used in MakeDataSet1D and MakeDataSet2D.
```
Comment (CumulonimbusCalvus) [inline src/qtt/data.py]: Is a rename of the argument `p` to something like 'parameter' possible?

File: src/qtt/data.py [hunk 4]
```diff
@@ -1029,183 +1031,326 @@ def saveExperimentData(outputdir, dataset, tag, dstr):
...
+
+        ValueError: When the shape of the measured data is not matching the shape of the setpoint array. When the
+                    measured data is a list, each list item must match the shape of the setpoint array.
+
+        TypeError: When a measurement name in the measurement list has an invalid type.
+
+    Returns:
+        The resulting data set and the measure names list
+    """
+    dd = new_data(arrays=(), location=location, loc_record=loc_record)
```
Comment (CumulonimbusCalvus) [inline src/qtt/data.py]: Can we change the variable `dd` to something more understandable?

File: src/qtt/data.py [hunk 5]
```diff
@@ -1029,183 +1031,326 @@ def saveExperimentData(outputdir, dataset, tag, dstr):
...
-    """ Determine unit associated with a parameter """
-    if isinstance(parameter, qcodes.Parameter):
-        return parameter.unit
+    if isinstance(zname, (str, qcodes.Parameter)):
+        measurement_list = [zname]
     else:
-        return None
+        measurement_list = zname
+
+    dd, _ = _make_data_set(measured_data_list, measurement_list, measurement_unit, location, loc_record,
```
Comment (CumulonimbusCalvus) [inline src/qtt/data.py]: `dd`?

File: src/qtt/data.py (hunk 1 above)
Comment (QFer) [reply src/qtt/data.py]: Removed it

File: src/qtt/data.py (hunk 2 above)
Comment (QFer) [reply src/qtt/data.py]: determine_parameter_unit was an existing method. I now added docstrings for it and added 100% unit tests for it.

File: src/qtt/data.py (hunk 3 above)
Comment (QFer) [reply src/qtt/data.py]: Absolutely!

File: src/qtt/data.py (hunk 4 above)
Comment (QFer) [reply src/qtt/data.py]: Yup, data_set it is...

File: src/qtt/data.py (hunk 5 above)
Comment (QFer) [reply src/qtt/data.py]: data_set it is...

File: src/tests/unittests/test_data.py [hunk 6]
```diff
@@ -154,3 +132,355 @@ def test_load_dataset(verbose=0):
...
+                               z, xunit='mV', yunit='Hz', zunit='A')
+
+    def test_dataset2dplain_shape_measuredata_list_ok(self):
+        v = [0, 1, 2.]
+        h = [0., 1, 2, 3, 4]
+        measurement = np.arange(len(v) * len(h)).reshape((len(v), len(h)))
+        z = [measurement, measurement]
+        data_set = qtt.data.makeDataSet2Dplain('horizontal', h, 'vertical', v, ['z1', 'z2'],
+                                               z, xunit='mV', yunit='Hz', zunit='A')
+        self.assertTrue(data_set.z1.shape == np.ones((len(v), len(h))).shape)
```
Comment (CumulonimbusCalvus) [inline src/tests/unittests/test_data.py]: General comment about testing equality of Numpy arrays. Is it sufficient to only check whether the shapes of the arrays are equal? I always check via:
```
desired_array = ???
self.assertTrue(np.array_equal(data_set.z1, desired_array))
```

Comment (QFer) [reply src/tests/unittests/test_data.py]: I changed where possible to the broader test with use of np.array_equal

**Pull request merged**